libmamba/src/solv-cpp/repo.cpp

```diff
@@ -83,34 +83,74 @@ namespace mamba::solv
 
     namespace
     {
-        struct FilePtr
+        class FilePtr
         {
-            std::FILE* ptr = nullptr;
-            std::string_view name = {};
+        public:
 
-            ~FilePtr() noexcept(false)
+            /**
+             * Open a file with C API.
+             *
+             * @param path must hae filesystem default encoding.
+             */
+            static auto open(std::filesystem::path path, const char* mode) -> FilePtr;
+
+            ~FilePtr() noexcept(false);
+
+            auto raw() noexcept -> std::FILE*;
+            auto raw() const noexcept -> const std::FILE*;
+
+        private:
+
+            FilePtr(std::FILE* ptr, std::string name);
+
+            std::FILE* m_ptr = nullptr;
+            std::string m_name = {};
+        };
+
+        FilePtr::FilePtr(std::FILE* ptr, std::string name)
+            : m_ptr{ ptr }
+            , m_name{ name }
+        {
+        }
+
+        auto FilePtr::open(std::filesystem::path path, const char* mode) -> FilePtr
+        {
+            std::string name = path;
+            std::FILE* ptr = std::fopen(name.c_str(), mode);
+            if (ptr == nullptr)
             {
-                const auto close_res = std::fclose(ptr);  // This flush too
-                if (close_res != 0)
-                {
-                    // TODO(C++20) fmt::format
-                    auto ss = std::stringstream();
-                    ss << "Unable to close file " << name;
-                    throw std::runtime_error(ss.str());
-                }
+                throw std::system_error(errno, std::generic_category());
```

Comment:
Why not throw filesystem error? I can see it was the previous code , but that makes little sense (I think it was like this previously because we ddidnt depend on the standard library filesystem)

Also, if this function throw, then it should be called in an opening constructor?

Reply:
> Why not throw filesystem error?

That would be better yes, but I this is the only way I found to make it work with the C error code...